autobahn/wamp/component.py

```diff
@@ -605,16 +605,20 @@ def transport_check(_):
                 transport_delay=delay,
             )
 
-            delay_f = txaio.sleep(delay)
-            txaio.add_callbacks(delay_f, attempt_connect, error)
+            self._delay_f = txaio.sleep(delay)
+            txaio.add_callbacks(self._delay_f, attempt_connect, error)
 
         # issue our first event, then start the reconnect loop
         start_f = self.fire('start', loop, self)
         txaio.add_callbacks(start_f, transport_check, error)
         return done_f
 
     def stop(self):
-        return self._session.leave()
+        if self._session.is_attached():
+            return self._session.leave()
+        else:
+            if getattr(self, '_delay_f', None):
+                return txaio.as_future(txaio.cancel, self._delay_f)
 
```

Comment:
You shouldn't need the `as_future` here I don't think .. just `txaio.cancel(..)`.

Also it's probably better to avoid the `getattr` call and just do `self._delay_f = None` in `__init__` or so? Also, it should be set back to `None` in `attempt_connect` I believe.

I haven't actually *tried* the code yet though :)

Reply:
Update the PR with the changes suggested above.

Reply:
Also how do we feel about a single method (Component.stop()) return different type of results. The user code would have to deal with a `Future` and None currently, thoughts ?

Reply:
Ah, yes missed that detail. Yes we should always return a Future/Deferred from `stop` .. so the `as_future` deals with that.

Reply:
Ok, brought that line back ;-)